### PR TITLE
[SYCL][NATIVECPU][LIBCLC] Implement missing builtins for half type

### PR DESCRIPTION
**Pull request by PietroGhg**
This PR implements missing builtins for `half` for the `generic` target and the missing overloads of `popcount` for `x86_64-unknown-linux`.
The accompanying UR change (https://github.com/oneapi-src/unified-runtime/pull/1628) reports `cl_khr_fp16` as supported, allowing us to correctly test the `half` overloads.

### CODE REVIEW

File: libclc/CMakeLists.txt [hunk 1]
```diff
@@ -42,6 +42,9 @@ set( LIBCLC_TARGETS_TO_BUILD "all"
 
 option( ENABLE_RUNTIME_SUBNORMAL "Enable runtime linking of subnormal support." OFF )
 
+set( LIBCLC_NATIVECPU_FLAGS_X86_64 ""
+  CACHE STRING "Semicolon-separated list on compiler flags for x86_64 libclc target.")
```
Comment (frasercrmck) [inline libclc/CMakeLists.txt:46]: typo: `of`.

Also, will this need updating for AMD in #13911? Should we prepare for that somehow?

File: libclc/x86_64-unknown-linux/libspirv/integer/popcount.cl [hunk 2]
```diff
@@ -1,3 +1,13 @@
-#include "helpers.h"
+#include <func.h>
+#include <clcmacro.h>
+#include <spirv/spirv.h>
 
-GEN_UNARY_BUILTIN(popcount)
+_CLC_DEFINE_UNARY_BUILTIN(int, __spirv_ocl_popcount, __builtin_popcount, int)
```
Comment (frasercrmck) [inline libclc/x86_64-unknown-linux/libspirv/integer/popcount.cl:5]: What's happening here?

**Review by frasercrmck — COMMENTED** [review]
How come some builtins are able to be given the `VECTORIZE` macro and others aren't?

**Comment by PietroGhg** [comment]
> How come some builtins are able to be given the `VECTORIZE` macro and others aren't?

`_CLC_DEFINE_UNARY_BUILTIN` and similar also call the `VECTORIZE` macro, if that's what you mean. For some builtins (e.g. `__spirv_ocl_frexp`), I couldn't use the `_CLC_DEFINE_ETC` macro, so the scalar builtin is defined in some way, and then we manually call the `VECTORIZE` macro. Hope this answers the question

File: libclc/x86_64-unknown-linux/libspirv/integer/popcount.cl (hunk 2 above)
Comment (PietroGhg) [reply libclc/x86_64-unknown-linux/libspirv/integer/popcount.cl:5]: The way `popcount` is defined in `generic` doesn't play well with the x86 ABI, so we were (and still are) over-riding it in this file. Previously we had our own macro that did what `_CLC_DEFINE_UNARY_BUILTIN` does (in `libclc/x86_64-unknown-linux/libspirv/integer/helpers.h`), so I've deleted it and decided to go with `_CLC_DEFINE_UNARY_BUILTIN`.

**Review by frasercrmck — APPROVED** [review]
LGTM

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 3]
```diff
@@ -129,8 +129,8 @@ if(SYCL_PI_UR_USE_FETCH_CONTENT)
   )
 
   fetch_adapter_source(native_cpu
-    ${UNIFIED_RUNTIME_REPO}
-    ${UNIFIED_RUNTIME_TAG}
+    "https://github.com/PietroGhg/unified-runtime.git"
+    pietro/native_cpu_half
```
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:133]:
```suggestion
    ${UNIFIED_RUNTIME_REPO}
    # commit 0e5bbcb2f8709643abf08cb83469787e866e3598
    # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
    # Date:   Wed Jun 19 14:33:26 2024 +0100
    #     Merge pull request #1628 from PietroGhg/pietro/native_cpu_half
    #     [NATIVECPU] Report cl_khr_fp16 on native cpu
    0e5bbcb2f8709643abf08cb83469787e866e3598
```

**Review by kbenzie — COMMENTED** [review]
https://github.com/oneapi-src/unified-runtime/pull/1628 has merged, please pull in the latest sycl branch changes and update the UR repo/tag as suggested.

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 3 above)
Comment (PietroGhg) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: That's done, thanks @kbenzie

**Review by kbenzie — APPROVED** [review]

**Comment by kbenzie** [comment]
@intel/llvm-gatekeepers please merge

**Pull request merged**